File: ios/Classes/YOLOMultiTaskView.swift
// Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

//  YOLOMultiTaskView — runs detect, segment, and classify on a single camera stream.
//  Three BasePredictor instances each receive raw CVPixelBuffers directly; no JPEG
//  round-trip or MethodChannel serialisation overhead. Each predictor uses the iOS
//  Neural Engine / GPU concurrently via Apple's CoreML async scheduling.

import AVFoundation
import CoreVideo
import UIKit
import UltralyticsYOLO

// MARK: - Per-predictor result adapter

/// Bridges ResultsListener / InferenceTimeListener callbacks back to a labelled slot
/// in YOLOMultiTaskView. All callbacks are re-dispatched onto `cameraQueue` so the
/// busy-flag bookkeeping stays on one serial queue (no locks needed).
final class MultiTaskPredictorAdapter: ResultsListener, InferenceTimeListener,
  @unchecked Sendable
{
  let taskName: String
  private let cameraQueue: DispatchQueue

  // Set by owner before use
  var onResult: ((YOLOResult) -> Void)?
  var onTime: ((Double, Double) -> Void)?

  init(taskName: String, cameraQueue: DispatchQueue) {
    self.taskName = taskName
    self.cameraQueue = cameraQueue
  }

  func on(result: YOLOResult) {
    let r = result
    cameraQueue.async { [weak self] in self?.onResult?(r) }
  }

  func on(inferenceTime: Double, fpsRate: Double) {
    let ms = inferenceTime
    let fps = fpsRate
    cameraQueue.async { [weak self] in self?.onTime?(ms, fps) }
  }
}

// MARK: - Detection box overlay

/// Lightweight UIView that draws detection bounding boxes directly over the camera preview.
/// Updated on the main thread after each detection result; redraws via setNeedsDisplay.
final class DetectionOverlayView: UIView {

  struct BoxInfo {
    let rect: CGRect
    let label: String
    let color: UIColor
  }

  var boxInfos: [BoxInfo] = [] { didSet { setNeedsDisplay() } }

  override init(frame: CGRect) {
    super.init(frame: frame)
    backgroundColor = .clear
    isUserInteractionEnabled = false
    isOpaque = false
  }

  required init?(coder: NSCoder) { fatalError() }

  override func draw(_ rect: CGRect) {
    guard let ctx = UIGraphicsGetCurrentContext() else { return }
    let fontSize: CGFloat = 11
    let font = UIFont.systemFont(ofSize: fontSize, weight: .semibold)
    let attrs: [NSAttributedString.Key: Any] = [
      .font: font,
      .foregroundColor: UIColor.white,
    ]

    for info in boxInfos {
      // Box stroke
      ctx.setStrokeColor(info.color.cgColor)
      ctx.setLineWidth(2)
      ctx.stroke(info.rect)

      // Label background + text
      let textSize = (info.label as NSString).size(withAttributes: attrs)
      let labelPad: CGFloat = 4
      let labelRect = CGRect(
        x: info.rect.minX,
        y: max(0, info.rect.minY - textSize.height - labelPad * 2),
        width: textSize.width + labelPad * 2,
        height: textSize.height + labelPad
      )
      ctx.setFillColor(info.color.withAlphaComponent(0.75).cgColor)
      ctx.fill(labelRect)
      (info.label as NSString).draw(
        at: CGPoint(x: labelRect.minX + labelPad, y: labelRect.minY + labelPad / 2),
        withAttributes: attrs
      )
    }
  }
}

// MARK: - YOLOMultiTaskView

/// A UIView that hosts a single AVCaptureSession and dispatches each incoming
/// camera frame to three independent YOLO predictors simultaneously.
@MainActor
public class YOLOMultiTaskView: UIView {

  // MARK: Camera

  private let captureSession = AVCaptureSession()
  private var previewLayer: AVCaptureVideoPreviewLayer?

  /// Serial queue for camera delegate callbacks and busy-flag mutations only.
  let cameraQueue = DispatchQueue(label: "yolo.multi-task.camera", qos: .userInteractive)

  /// Per-predictor inference queues. predict() is synchronous (blocks the caller), so each
  /// predictor runs on its own queue so all three run concurrently instead of serially.
  private let detectQueue = DispatchQueue(label: "yolo.infer.detect", qos: .userInteractive)
  private let segmentQueue = DispatchQueue(label: "yolo.infer.segment", qos: .userInteractive)
  private let classifyQueue = DispatchQueue(label: "yolo.infer.classify", qos: .userInteractive)

  // MARK: Predictors

  var detectPredictor: BasePredictor?
  var segmentPredictor: BasePredictor?
  var classifyPredictor: BasePredictor?

  /// One-frame-deep back-pressure per predictor: if the previous frame is still
  /// being processed we skip rather than queue up. Accessed only on cameraQueue.
  var detectBusy = false
  var segmentBusy = false
  var classifyBusy = false

  private lazy var detectAdapter = MultiTaskPredictorAdapter(taskName: "detect", cameraQueue: cameraQueue)
  private lazy var segmentAdapter = MultiTaskPredictorAdapter(taskName: "segment", cameraQueue: cameraQueue)
  private lazy var classifyAdapter = MultiTaskPredictorAdapter(taskName: "classify", cameraQueue: cameraQueue)

  // MARK: Per-task FPS tracking (cameraQueue)

  private var detectLastResultTime: Double = 0
  private var segmentLastResultTime: Double = 0
  private var classifyLastResultTime: Double = 0

  private var detectFps: Double = 0
  private var segmentFps: Double = 0
  private var classifyFps: Double = 0

  // Camera FPS (cameraQueue)
  private var camFrameCount = 0
  private var camFpsWindowStart: Double = 0
  private var camFps: Double = 0

  // MARK: Callback

  /// Called on the main thread with a stream-data dict. Keys: "type" (detect/segment/classify),
  /// "detections", "processingTimeMs", "fps", "cameraFps".
  var onMultiTaskStream: (([String: Any]) -> Void)?

  // MARK: Detection overlay

  private let detectOverlay = DetectionOverlayView()

  // MARK: Loading indicator

  public let activityIndicator = UIActivityIndicatorView(style: .large)
  private var loadedCount = 0
  private var expectedCount = 0

  // MARK: Init

  public override init(frame: CGRect) {
    super.init(frame: frame)
    setupUI()
    wireAdapters()
  }

  required init?(coder: NSCoder) {
    super.init(coder: coder)
    setupUI()
    wireAdapters()
  }

  private func setupUI() {
    backgroundColor = .black
    addSubview(detectOverlay)
    activityIndicator.color = .white
    activityIndicator.startAnimating()
    addSubview(activityIndicator)
  }

  public override func layoutSubviews() {
    super.layoutSubviews()
    previewLayer?.frame = bounds
    detectOverlay.frame = bounds
    activityIndicator.center = CGPoint(x: bounds.midX, y: bounds.midY)
  }

  // MARK: - Adapter wiring

  private func wireAdapters() {
    detectAdapter.onResult = { [weak self] result in self?.handleResult(result, task: "detect") }
    segmentAdapter.onResult = { [weak self] result in self?.handleResult(result, task: "segment") }
    classifyAdapter.onResult = { [weak self] result in self?.handleResult(result, task: "classify") }

    detectAdapter.onTime = { [weak self] _, fps in self?.detectFps = fps }
    segmentAdapter.onTime = { [weak self] _, fps in self?.segmentFps = fps }
    classifyAdapter.onTime = { [weak self] _, fps in self?.classifyFps = fps }
  }

  // MARK: - Result handling (cameraQueue)

  private func handleResult(_ result: YOLOResult, task: String) {
    // Clear busy flag so the next frame can be dispatched.
    switch task {
    case "detect":
      detectBusy = false
      detectPredictor?.isUpdating = false
    case "segment":
      segmentBusy = false
      segmentPredictor?.isUpdating = false
    default:
      classifyBusy = false
      classifyPredictor?.isUpdating = false
    }

    // FPS from result interval
    let now = CACurrentMediaTime()
    var taskFps: Double = 0
    switch task {
    case "detect":
      if detectLastResultTime > 0 {
        let dt = now - detectLastResultTime
        if dt > 0 { detectFps = 1.0 / dt }
      }
      detectLastResultTime = now
      taskFps = detectFps
    case "segment":
      if segmentLastResultTime > 0 {
        let dt = now - segmentLastResultTime
        if dt > 0 { segmentFps = 1.0 / dt }
      }
      segmentLastResultTime = now
      taskFps = segmentFps
    default:
      if classifyLastResultTime > 0 {
        let dt = now - classifyLastResultTime
        if dt > 0 { classifyFps = 1.0 / dt }
      }
      classifyLastResultTime = now
      taskFps = classifyFps
    }

    let camFpsSnapshot = camFps
    let streamData = buildStreamData(result: result, task: task, fps: taskFps, cameraFps: camFpsSnapshot)

    // Snapshot detection data for the native overlay; rect computation deferred to main thread
    // so we have the correct view bounds.
    let rawBoxes: [(xywhn: CGRect, label: String)] = task == "detect"
      ? result.boxes.prefix(30).map { ($0.xywhn, "\($0.cls) \(Int($0.conf * 100))%") }
      : []
    let origShape = result.orig_shape

    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      if task == "detect" {
        let viewSize = self.bounds.size
        self.detectOverlay.boxInfos = rawBoxes.map { b in
          DetectionOverlayView.BoxInfo(
            rect: self.aspectFillRect(normalized: b.xywhn, imageSize: origShape, viewSize: viewSize),
            label: b.label,
            color: .systemBlue
          )
        }
      }
      self.onMultiTaskStream?(streamData)
    }
  }

  // MARK: - Coordinate mapping

  /// Maps a normalized bounding box (0–1 in image space) to screen points, accounting for
  /// the AVCaptureVideoPreviewLayer's resizeAspectFill scaling — identical to YOLOView's
  /// `aspectFillDisplayRect` helper in the UltralyticsYOLO framework.
  private func aspectFillRect(normalized: CGRect, imageSize: CGSize, viewSize: CGSize) -> CGRect {
    guard imageSize.width > 0, imageSize.height > 0,
      viewSize.width > 0, viewSize.height > 0
    else { return .zero }
    let scale = max(viewSize.width / imageSize.width, viewSize.height / imageSize.height)
    let offsetX = (imageSize.width * scale - viewSize.width) / 2
    let offsetY = (imageSize.height * scale - viewSize.height) / 2
    return CGRect(
      x: normalized.minX * imageSize.width * scale - offsetX,
      y: normalized.minY * imageSize.height * scale - offsetY,
      width: normalized.width * imageSize.width * scale,
      height: normalized.height * imageSize.height * scale
    )
  }

  // MARK: - Stream data builder

  private func buildStreamData(
    result: YOLOResult, task: String, fps: Double, cameraFps: Double
  ) -> [String: Any] {
    var map: [String: Any] = [
      "type": task,
      "fps": fps,
      "cameraFps": cameraFps,
      "processingTimeMs": result.speed * 1000,
    ]

    switch task {
    case "classify":
      if let probs = result.probs {
        var top5: [[String: Any]] = []
        for i in 0..<min(probs.top5Labels.count, probs.top5Confs.count) {
          top5.append(["name": probs.top5Labels[i], "confidence": Double(probs.top5Confs[i])])
        }
        map["classification"] = [
          "top1": probs.top1Label,
          "top1Confidence": Double(probs.top1Conf),
          "top5": top5,
        ]
      }
    default:
      var detections: [[String: Any]] = []
      for box in result.boxes.prefix(50) {
        detections.append([
          "className": box.cls,
          "confidence": Double(box.conf),
          "normalizedBox": [
            "left": Double(box.xywhn.minX),
            "top": Double(box.xywhn.minY),
            "right": Double(box.xywhn.maxX),
            "bottom": Double(box.xywhn.maxY),
          ],
        ])
      }
      map["detections"] = detections
    }

    return map
  }

  // MARK: - Model loading

  /// Load all three models. `completion` fires on the main thread once all three finish
  /// (or fail silently with a log). Camera starts automatically after all models are ready.
  func loadModels(
    detectPath: String,
    segmentPath: String,
    classifyPath: String,
    useGpu: Bool = true,
    cameraPosition: AVCaptureDevice.Position = .back,
    completion: @escaping () -> Void
  ) {
    expectedCount = 3
    loadedCount = 0

    func tryDone() {
      loadedCount += 1
      if loadedCount == expectedCount {
        DispatchQueue.main.async { [weak self] in
          self?.activityIndicator.stopAnimating()
          self?.startCamera(position: cameraPosition)
          completion()
        }
      }
    }

    load(path: detectPath, task: .detect, useGpu: useGpu) { [weak self] p in
      self?.detectPredictor = p
      tryDone()
    }
    load(path: segmentPath, task: .segment, useGpu: useGpu) { [weak self] p in
      self?.segmentPredictor = p
      tryDone()
    }
    load(path: classifyPath, task: .classify, useGpu: useGpu) { [weak self] p in
      self?.classifyPredictor = p
      tryDone()
    }
  }

  private func load(
    path: String, task: YOLOTask, useGpu: Bool,
    completion: @escaping (BasePredictor?) -> Void
  ) {
    guard let url = resolveModelURL(path) else {
      NSLog("YOLOMultiTaskView: model not found: %@", path)
      completion(nil)
      return
    }
    BasePredictor.create(for: task, modelURL: url, isRealTime: true, useGpu: useGpu) { result in
      switch result {
      case .success(let p):
        let bp = p as? BasePredictor
        bp?.capturesOriginalImage = false
        completion(bp)
      case .failure(let err):
        NSLog("YOLOMultiTaskView: load failed for %@: %@", path, err.localizedDescription)
        completion(nil)
      }
    }
  }

  private func resolveModelURL(_ nameOrPath: String) -> URL? {
    let lc = nameOrPath.lowercased()
    if lc.hasSuffix(".mlmodelc") || lc.hasSuffix(".mlpackage") || lc.hasSuffix(".mlmodel") {
      let u = URL(fileURLWithPath: nameOrPath)
      if FileManager.default.fileExists(atPath: u.path) { return u }
    }
    if let u = Bundle.main.url(forResource: nameOrPath, withExtension: "mlmodelc") { return u }
    if let u = Bundle.main.url(forResource: nameOrPath, withExtension: "mlpackage") { return u }
    return nil
  }

  // MARK: - Camera

  private func startCamera(position: AVCaptureDevice.Position) {
    let pos = position
    cameraQueue.async { [weak self] in self?.setupCamera(position: pos) }
  }

  private func setupCamera(position: AVCaptureDevice.Position) {
    captureSession.beginConfiguration()
    captureSession.sessionPreset = .hd1280x720

    guard let device = bestCaptureDevice(position: position),
      let input = try? AVCaptureDeviceInput(device: device),
      captureSession.canAddInput(input)
    else {
      captureSession.commitConfiguration()
      return
    }
    captureSession.addInput(input)

    let output = AVCaptureVideoDataOutput()
    output.videoSettings = [
      kCVPixelBufferPixelFormatTypeKey as String: NSNumber(value: kCVPixelFormatType_32BGRA)
    ]
    output.alwaysDiscardsLateVideoFrames = true
    output.setSampleBufferDelegate(self, queue: cameraQueue)
    if captureSession.canAddOutput(output) { captureSession.addOutput(output) }

    if let conn = output.connection(with: .video) {
      conn.videoOrientation = .portrait
    }

    captureSession.commitConfiguration()

    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      let preview = AVCaptureVideoPreviewLayer(session: self.captureSession)
      preview.videoGravity = .resizeAspectFill
      preview.frame = self.bounds
      self.layer.insertSublayer(preview, at: 0)
      self.previewLayer = preview
    }

    captureSession.startRunning()
    camFpsWindowStart = CACurrentMediaTime()
  }

  public func stopCamera() {
    cameraQueue.async { [weak self] in
      self?.captureSession.stopRunning()
    }
  }

  deinit {
    if captureSession.isRunning {
      captureSession.stopRunning()
    }
  }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

extension YOLOMultiTaskView: AVCaptureVideoDataOutputSampleBufferDelegate, @unchecked Sendable {

  public func captureOutput(
    _ output: AVCaptureOutput,
    didOutput sampleBuffer: CMSampleBuffer,
    from connection: AVCaptureConnection
  ) {
    // All mutable state accesses in this extension run on cameraQueue.

    // Track camera FPS
    let now = CACurrentMediaTime()
    camFrameCount += 1
    let elapsed = now - camFpsWindowStart
    if elapsed >= 0.5 {
      camFps = Double(camFrameCount) / elapsed
      camFrameCount = 0
      camFpsWindowStart = now
    }

    // Dispatch each predictor to its own queue so all three run concurrently.
    // predict() is synchronous — calling it on cameraQueue would serialise all three
    // and block new frames from arriving. Capture the adapters here (on cameraQueue)
    // so there is no cross-queue access to lazy vars.
    if let p = detectPredictor, !detectBusy, !p.isUpdating {
      detectBusy = true
      p.isUpdating = true
      let buf = sampleBuffer
      let adapter = detectAdapter
      detectQueue.async { p.predict(sampleBuffer: buf, onResultsListener: adapter, onInferenceTime: adapter) }
    }
    if let p = segmentPredictor, !segmentBusy, !p.isUpdating {
      segmentBusy = true
      p.isUpdating = true
      let buf = sampleBuffer
      let adapter = segmentAdapter
      segmentQueue.async { p.predict(sampleBuffer: buf, onResultsListener: adapter, onInferenceTime: adapter) }
    }
    if let p = classifyPredictor, !classifyBusy, !p.isUpdating {
      classifyBusy = true
      p.isUpdating = true
      let buf = sampleBuffer
      let adapter = classifyAdapter
      classifyQueue.async { p.predict(sampleBuffer: buf, onResultsListener: adapter, onInferenceTime: adapter) }
    }
  }
}

File: android/src/main/kotlin/com/ultralytics/yolo/YOLOMultiTaskView.kt
// Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

package com.ultralytics.yolo

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.ProgressBar
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

// MARK: - Detection box overlay

class DetectionOverlayView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : View(context, attrs) {

    data class BoxInfo(val rect: RectF, val label: String, val color: Int)

    private var boxes: List<BoxInfo> = emptyList()

    private val boxPaint = Paint().apply {
        style = Paint.Style.STROKE
        strokeWidth = 4f
        isAntiAlias = true
    }
    private val fillPaint = Paint().apply {
        style = Paint.Style.FILL
        isAntiAlias = true
    }
    private val textPaint = Paint().apply {
        color = Color.WHITE
        textSize = 28f
        isAntiAlias = true
        isFakeBoldText = true
    }

    init {
        setWillNotDraw(false)
        setBackgroundColor(Color.TRANSPARENT)
    }

    fun setDetections(newBoxes: List<BoxInfo>) {
        boxes = newBoxes
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val labelPad = 6f
        for (info in boxes) {
            boxPaint.color = info.color
            canvas.drawRect(info.rect, boxPaint)

            val textWidth = textPaint.measureText(info.label)
            val textHeight = textPaint.descent() - textPaint.ascent()
            val labelTop = maxOf(0f, info.rect.top - textHeight - labelPad * 2)
            val labelRect = RectF(
                info.rect.left,
                labelTop,
                info.rect.left + textWidth + labelPad * 2,
                labelTop + textHeight + labelPad
            )
            fillPaint.color = Color.argb(192, Color.red(info.color), Color.green(info.color), Color.blue(info.color))
            canvas.drawRect(labelRect, fillPaint)
            canvas.drawText(
                info.label,
                labelRect.left + labelPad,
                labelRect.bottom - labelPad / 2 - textPaint.descent(),
                textPaint
            )
        }
    }
}

// MARK: - YOLOMultiTaskView

/**
 * Android equivalent of the iOS YOLOMultiTaskView. Hosts a single CameraX session and
 * dispatches each camera frame to three independent YOLO predictors (detect, segment,
 * classify) concurrently on their own background threads. Results are delivered to
 * [onMultiTaskStream] on the main thread.
 */
class YOLOMultiTaskView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : FrameLayout(context, attrs) {

    companion object {
        private const val TAG = "YOLOMultiTaskView"
    }

    // Camera
    private val previewView = PreviewView(context)
    private var cameraProvider: ProcessCameraProvider? = null
    private var cameraExecutor: ExecutorService? = null

    // Predictors
    @Volatile private var detectPredictor: Predictor? = null
    @Volatile private var segmentPredictor: Predictor? = null
    @Volatile private var classifyPredictor: Predictor? = null

    // Per-predictor executors — each runs predict() concurrently
    private val detectExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private val segmentExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private val classifyExecutor: ExecutorService = Executors.newSingleThreadExecutor()

    // One-frame-deep back-pressure flags (atomic so they're safe to read on cameraExecutor)
    private val detectBusy = AtomicBoolean(false)
    private val segmentBusy = AtomicBoolean(false)
    private val classifyBusy = AtomicBoolean(false)

    // Per-task FPS tracking (only accessed on their respective executor threads)
    private var detectLastResultTime: Long = 0
    private var segmentLastResultTime: Long = 0
    private var classifyLastResultTime: Long = 0

    // Camera FPS (accessed on cameraExecutor)
    private var camFrameCount = 0
    private var camFpsWindowStart: Long = System.currentTimeMillis()
    @Volatile private var camFps: Double = 0.0

    // Views
    private val overlayView = DetectionOverlayView(context)
    private val progressBar = ProgressBar(context)

    private var lifecycleOwner: LifecycleOwner? = null

    /** Called on the main thread for each inference result. */
    var onMultiTaskStream: ((Map<String, Any>) -> Unit)? = null

    init {
        setBackgroundColor(Color.BLACK)
        addView(previewView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        addView(overlayView, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        val pbParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        pbParams.gravity = Gravity.CENTER
        addView(progressBar, pbParams)
    }

    fun onLifecycleOwnerAvailable(owner: LifecycleOwner) {
        lifecycleOwner = owner
    }

    // MARK: - Model loading

    fun loadModels(
        detectPath: String,
        segmentPath: String,
        classifyPath: String,
        useGpu: Boolean,
        lensFacing: Int = CameraSelector.LENS_FACING_BACK,
        onLoaded: (() -> Unit)? = null
    ) {
        var loadedCount = 0

        fun onOneLoaded() {
            val allDone = synchronized(this) {
                loadedCount++
                loadedCount == 3
            }
            if (allDone) {
                post {
                    progressBar.visibility = View.GONE
                    startCamera(lensFacing)
                    onLoaded?.invoke()
                }
            }
        }

        Executors.newSingleThreadExecutor().execute {
            try {
                detectPredictor = ObjectDetector(context = context, modelPath = detectPath, labels = emptyList(), useGpu = useGpu)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load detect model: $detectPath", e)
            }
            onOneLoaded()
        }

        Executors.newSingleThreadExecutor().execute {
            try {
                segmentPredictor = Segmenter(context, segmentPath, labels = emptyList(), useGpu = useGpu)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load segment model: $segmentPath", e)
            }
            onOneLoaded()
        }

        Executors.newSingleThreadExecutor().execute {
            try {
                classifyPredictor = Classifier(context, classifyPath, labels = emptyList(), useGpu = useGpu)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load classify model: $classifyPath", e)
            }
            onOneLoaded()
        }
    }

    // MARK: - Camera

    private fun startCamera(lensFacing: Int) {
        val owner = lifecycleOwner ?: return
        val future = ProcessCameraProvider.getInstance(context)
        future.addListener({
            cameraProvider = future.get()
            bindCamera(lensFacing, owner)
        }, ContextCompat.getMainExecutor(context))
    }

    private fun bindCamera(lensFacing: Int, owner: LifecycleOwner) {
        val provider = cameraProvider ?: return

        val cameraSelector = CameraSelector.Builder()
            .requireLensFacing(lensFacing)
            .build()

        val preview = Preview.Builder()
            .setTargetAspectRatio(AspectRatio.RATIO_4_3)
            .build()
        preview.setSurfaceProvider(previewView.surfaceProvider)

        val analysis = ImageAnalysis.Builder()
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .setTargetAspectRatio(AspectRatio.RATIO_4_3)
            .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
            .build()

        val exec = Executors.newSingleThreadExecutor()
        cameraExecutor = exec
        analysis.setAnalyzer(exec) { imageProxy -> onFrame(imageProxy) }

        try {
            provider.unbindAll()
            provider.bindToLifecycle(owner, cameraSelector, preview, analysis)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to bind camera", e)
        }
    }

    // MARK: - Frame processing

    private fun onFrame(imageProxy: ImageProxy) {
        // Track camera FPS
        camFrameCount++
        val now = System.currentTimeMillis()
        val elapsed = now - camFpsWindowStart
        if (elapsed >= 500) {
            camFps = camFrameCount * 1000.0 / elapsed
            camFrameCount = 0
            camFpsWindowStart = now
        }

        val rotationDegrees = imageProxy.imageInfo.rotationDegrees
        val isRotated = rotationDegrees % 180 != 0
        val w = imageProxy.width
        val h = imageProxy.height
        val orientedWidth = if (isRotated) h else w
        val orientedHeight = if (isRotated) w else h

        val bitmap = ImageUtils.toBitmap(imageProxy)
        imageProxy.close()
        if (bitmap == null) return

        val camFpsNow = camFps

        // Detect
        if (!detectBusy.getAndSet(true)) {
            val p = detectPredictor
            if (p != null) {
                detectExecutor.execute {
                    try {
                        val t0 = System.nanoTime()
                        val result = p.predict(bitmap, orientedWidth, orientedHeight, rotateForCamera = true)
                        val ms = (System.nanoTime() - t0) / 1_000_000.0
                        val nowMs = System.currentTimeMillis()
                        val fps = if (detectLastResultTime > 0L) 1000.0 / (nowMs - detectLastResultTime) else 0.0
                        detectLastResultTime = nowMs

                        val data = buildStreamData("detect", result, ms, fps, camFpsNow)
                        post {
                            val imgW = result.origShape.width.toFloat()
                            val imgH = result.origShape.height.toFloat()
                            val vW = width.toFloat()
                            val vH = height.toFloat()
                            overlayView.setDetections(result.boxes.take(30).map { box ->
                                DetectionOverlayView.BoxInfo(
                                    rect = aspectFillRect(box.xywhn, imgW, imgH, vW, vH),
                                    label = "${box.cls} ${(box.conf * 100).toInt()}%",
                                    color = Color.rgb(59, 130, 246)
                                )
                            })
                            onMultiTaskStream?.invoke(data)
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Detect inference error", e)
                    } finally {
                        detectBusy.set(false)
                    }
                }
            } else {
                detectBusy.set(false)
            }
        }

        // Segment
        if (!segmentBusy.getAndSet(true)) {
            val p = segmentPredictor
            if (p != null) {
                segmentExecutor.execute {
                    try {
                        val t0 = System.nanoTime()
                        val result = p.predict(bitmap, orientedWidth, orientedHeight, rotateForCamera = true)
                        val ms = (System.nanoTime() - t0) / 1_000_000.0
                        val nowMs = System.currentTimeMillis()
                        val fps = if (segmentLastResultTime > 0L) 1000.0 / (nowMs - segmentLastResultTime) else 0.0
                        segmentLastResultTime = nowMs

                        val data = buildStreamData("segment", result, ms, fps, camFpsNow)
                        post { onMultiTaskStream?.invoke(data) }
                    } catch (e: Exception) {
                        Log.e(TAG, "Segment inference error", e)
                    } finally {
                        segmentBusy.set(false)
                    }
                }
            } else {
                segmentBusy.set(false)
            }
        }

        // Classify
        if (!classifyBusy.getAndSet(true)) {
            val p = classifyPredictor
            if (p != null) {
                classifyExecutor.execute {
                    try {
                        val t0 = System.nanoTime()
                        val result = p.predict(bitmap, orientedWidth, orientedHeight, rotateForCamera = true)
                        val ms = (System.nanoTime() - t0) / 1_000_000.0
                        val nowMs = System.currentTimeMillis()
                        val fps = if (classifyLastResultTime > 0L) 1000.0 / (nowMs - classifyLastResultTime) else 0.0
                        classifyLastResultTime = nowMs

                        val data = buildStreamData("classify", result, ms, fps, camFpsNow)
                        post { onMultiTaskStream?.invoke(data) }
                    } catch (e: Exception) {
                        Log.e(TAG, "Classify inference error", e)
                    } finally {
                        classifyBusy.set(false)
                    }
                }
            } else {
                classifyBusy.set(false)
            }
        }
    }

    // MARK: - Coordinate mapping

    /** Maps a normalized box (0-1) to screen pixels under resizeAspectFill. */
    private fun aspectFillRect(normalized: RectF, imgW: Float, imgH: Float, viewW: Float, viewH: Float): RectF {
        if (imgW <= 0 || imgH <= 0 || viewW <= 0 || viewH <= 0) return RectF()
        val scale = maxOf(viewW / imgW, viewH / imgH)
        val offsetX = (imgW * scale - viewW) / 2f
        val offsetY = (imgH * scale - viewH) / 2f
        return RectF(
            normalized.left * imgW * scale - offsetX,
            normalized.top * imgH * scale - offsetY,
            normalized.right * imgW * scale - offsetX,
            normalized.bottom * imgH * scale - offsetY
        )
    }

    // MARK: - Stream data builder

    private fun buildStreamData(
        task: String,
        result: YOLOResult,
        processingMs: Double,
        fps: Double,
        cameraFps: Double
    ): Map<String, Any> {
        val map = HashMap<String, Any>()
        map["type"] = task
        map["fps"] = fps
        map["cameraFps"] = cameraFps
        map["processingTimeMs"] = processingMs

        when (task) {
            "classify" -> {
                result.probs?.let { probs ->
                    val top5 = ArrayList<Map<String, Any>>()
                    for (i in 0 until minOf(probs.top5Labels.size, probs.top5Confs.size)) {
                        top5.add(mapOf(
                            "name" to probs.top5Labels[i],
                            "confidence" to probs.top5Confs[i].toDouble()
                        ))
                    }
                    map["classification"] = mapOf(
                        "top1" to probs.top1Label,
                        "top1Confidence" to probs.top1Conf.toDouble(),
                        "top5" to top5
                    )
                }
            }
            else -> {
                val detections = ArrayList<Map<String, Any>>()
                for (box in result.boxes.take(50)) {
                    detections.add(mapOf(
                        "className" to box.cls,
                        "confidence" to box.conf.toDouble(),
                        "normalizedBox" to mapOf(
                            "left" to box.xywhn.left.toDouble(),
                            "top" to box.xywhn.top.toDouble(),
                            "right" to box.xywhn.right.toDouble(),
                            "bottom" to box.xywhn.bottom.toDouble()
                        )
                    ))
                }
                map["detections"] = detections
            }
        }
        return map
    }

    // MARK: - Lifecycle

    fun stopCamera() {
        cameraProvider?.unbindAll()
        cameraExecutor?.shutdown()
        cameraExecutor = null
    }

    fun release() {
        stopCamera()
        detectExecutor.shutdown()
        segmentExecutor.shutdown()
        classifyExecutor.shutdown()
        (detectPredictor as? BasePredictor)?.close()
        (segmentPredictor as? BasePredictor)?.close()
        (classifyPredictor as? BasePredictor)?.close()
        detectPredictor = null
        segmentPredictor = null
        classifyPredictor = null
    }
}
